services/spar/src/Spar/App.hs

```diff
@@ -186,47 +200,70 @@ instance Intra.MonadSparToBrig Spar where
   -- the InResponseTo attribute MUST match the request's ID.
   reqid <- either (throwSpar . SparNoRequestRefInResponse . cs) pure $ SAML.rspInResponseTo aresp
   format :: Maybe VerdictFormat <- wrapMonadClient $ Data.getVerdictFormat reqid
   case format of
-    Just (VerdictFormatWeb) -> verdictHandlerWeb verdict
-    Just (VerdictFormatMobile granted denied) -> verdictHandlerMobile granted denied verdict
+    Just (VerdictFormatWeb)
+      -> verdictHandlerResult cky verdict >>= verdictHandlerWeb
+    Just (VerdictFormatMobile granted denied)
+      -> verdictHandlerResult cky verdict >>= verdictHandlerMobile granted denied
     Nothing -> throwError $ SAML.BadSamlResponse "AuthRequest seems to have disappeared (could not find verdict format)."
                -- (this shouldn't happen too often, see 'storeVerdictFormat')
 
 data VerdictHandlerResult
   = VerifyHandlerDenied [ST]
   | VerifyHandlerGranted SetCookie UserId
 
-verdictHandlerResult :: HasCallStack => SAML.AccessVerdict -> Spar VerdictHandlerResult
-verdictHandlerResult = \case
+verdictHandlerResult :: HasCallStack => Maybe BindCookie -> SAML.AccessVerdict -> Spar VerdictHandlerResult
+verdictHandlerResult bindCky = \case
   denied@(SAML.AccessDenied reasons) -> do
     SAML.logger SAML.Debug (show denied)
     pure $ VerifyHandlerDenied reasons
-  SAML.AccessGranted userref -> do
-    uid :: UserId    <- maybe (createUser userref) pure =<< getUser userref
-    cky :: SetCookie <- Intra.ssoLogin uid  -- TODO: can this be a race condition?  (user is not
-                                            -- quite created yet when we ask for a cookie?  do we do
-                                            -- quorum reads / writes here?  writes: probably yes,
-                                            -- reads: probably no.)
+
+  granted@(SAML.AccessGranted userref) -> do
+    uid :: UserId <- do
+      SAML.logger SAML.Debug (show granted)
+      fromBindCookie <- maybe (pure Nothing) (wrapMonadClient . Data.lookupBindCookie) bindCky
+      fromSparCass   <- getUser userref
+        -- race conditions: if the user has been created on spar, but not on brig, 'getUser'
+        -- returns 'Nothing'.  this is ok assuming 'createUser', 'bindUser' (called below) are
+        -- idempotent.
+
+      case (fromBindCookie, fromSparCass) of
+        (Nothing,  Nothing)   -> createUser userref      -- first sso authentication
+        (Nothing,  Just uid)  -> pure uid                -- sso re-authentication
+        (Just uid, Nothing)   -> bindUser uid userref    -- bind existing user (non-sso or sso) to ssoid
+        (Just uid, Just uid')
+          | uid == uid' -> pure uid                      -- redundant binding (no change to brig or spar)
+          | otherwise -> throwSpar SparBindUserRefTaken  -- attempt to use ssoid for a second wire user
+
+    cky :: SetCookie <- retrySparAction $ Intra.ssoLogin uid
```

Comment:
This should not be needed, given that there is a write Quorum in `createUser`; indeed it may be harmless but perhaps hiding other assumptions?